lib/profileAgents.js

```diff
@@ -0,0 +1,361 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+const assert = require('assert-plus');
+const base64url = require('base64url-universal');
+const bedrock = require('bedrock');
+const crypto = require('crypto');
+const {CapabilityAgent} = require('webkms-client');
+const database = require('bedrock-mongodb');
+const {promisify} = require('util');
+const kms = require('./kms');
+const utils = require('./utils');
+const zcaps = require('./zcaps');
+
+const {util: {BedrockError}} = bedrock;
+
+// load config defaults
+require('./config');
+const KMS_MODULE = 'ssm-v1';
+
+bedrock.events.on('bedrock-mongodb.ready', async () => {
+  await promisify(database.openCollections)(['profile-profileAgent']);
+
+  await promisify(database.createIndexes)([{
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.id': 1},
+    options: {unique: true, background: false}
+  }, {
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.profile': 1},
+    options: {unique: false, background: false}
+  }, {
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.controller': 1},
+    options: {unique: true, background: false}
+  }, {
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.id': 1, 'profileAgent.sequence': 1},
+    options: {unique: false, background: false}
+  }, {
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.account': 1, 'profileAgent.id': 1},
+    options: {
+      partialFilterExpression: {account: {$exists: true}},
+      unique: true,
+      background: false
+    }
+  }]);
+});
+
+/**
+ * Creates a ProfileAgent and associates it with an account if provided.
+ *
+ * @param {String} accountId the id of the account to associate with the
+ *                           ProfileAgent.
+ *
+ * @return {Promise<ProfileAgent>} resolves to a ProfileAgent.
+ */
+exports.create = async ({accountId} = {}) => {
+  // 1. Generate a random secret
+  const secret = base64url.encode(crypto.randomBytes(32));
+  const handle = 'primary';
+  // 2. Generate capability agent for the zCap key
+  const capabilityAgent = await CapabilityAgent.fromSecret({handle, secret});
+  // 3. Create keystore in order to create the zCap key for the profile agent
+  // TODO: Will want to store the capability agent up here first, otherwise we
+  //       have no clean way of cleaning up the other stuff that gets created
+  //       on other systems should we experience a failure.
+  const keystore = await kms.createKeystore(
+    {capabilityAgent, referenceId: 'primary'});
+  // 4. Create the zCap key for the profile agent
+  const keystoreAgent = kms.getKeystoreAgent({capabilityAgent, keystore});
+  const key = await keystoreAgent.generateKey(
+    {type: 'Ed25519VerificationKey2018', kmsModule: KMS_MODULE});
+  // 5. Use the zCap key to create the DID for the profile agent
+  await utils.setKeyId({key});
+  // 6. Generate capability agent for ProfileAgent
```

Comment:
This language is confusing since this capability agent is one representation of the profile agent. Note that it's `id` matches that of the profile agent. You use this agent interface to invoke zcaps on behalf of the profile. Perhaps:

```suggestion
  // 6. Generate profile agent's CapabilityAgent instance
```